tests/tracer/test_single_span_sampling_rule.py

```diff
@@ -0,0 +1,149 @@
+
+
+def test_single_span_rule_match_only_service():
+    rule = SpanSamplingRule(service="test_service")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+def test_single_span_rule_match_only_name():
+    rule = SpanSamplingRule(name="test_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+# More extensive testing of the matching is done in test_glob_matcher.py
+def test_single_span_rule_match_wildcards():
+    rule = SpanSamplingRule(service="test_*", name="test?????")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+def test_single_span_rule_no_match_wildcards():
+    rule = SpanSamplingRule(service="*test_", name="test_nam??")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_unsupported_pattern_bracket_expansion():
+    rule = SpanSamplingRule(service="test_servic[a-z]+", name="test_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_unsupported_pattern_escape_character():
+    rule = SpanSamplingRule(service="test_servic[?]", name="test_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_unsupported_pattern_bracket_expansion_literal_evaluation():
+    rule = SpanSamplingRule(service="test_servic[a-z]+", name="test_name")
+    span = traced_function(rule, service="test_servic[a-z]+")
+
+    assert_sampling_decision_tags(span)
+
+
+def test_single_span_rule_unsupported_pattern_escape_character_literal_evaluation():
+    rule = SpanSamplingRule(service="test_servic[?]", name="test_name")
+    span = traced_function(rule, service="test_servic[?]")
+
+    assert_sampling_decision_tags(span)
+
+
+def test_multiple_span_rule_match():
+    rule = SpanSamplingRule(service="test_service", name="test_name")
+    for i in range(10):
+        span = traced_function(rule)
+
+        assert_sampling_decision_tags(span)
+
+
+def test_rate_limiter_0():
+    rule = SpanSamplingRule(service="test_service", name="test_name", max_per_second=0)
+    for i in range(10):
+        span = traced_function(rule)
+        # None of the single span sampling tags should be added to the span
+        # if it's dropped by the single span rate limiter
+        assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_match_max_per_sec():
+    rule = SpanSamplingRule(service="test_service", name="test_name", max_per_second=2)
+    # Make spans till we hit the limit, then make a span while the limit is hit and make sure tags were not added.
+    while True:
+        span = traced_function(rule)
+        if not rule.limiter.is_allowed(span.start_ns):
+            break
+
+    span = traced_function(rule)
```

Comment:
If I'm understanding correctly, the call here to `traced_function` could occur after enough time such that the rate limiter would allow the span sampling rule to match.

Reply:
I think it's pretty unlikely and have run the test a number of times, but it coullld happen. It should be nanoseconds or less between us hitting the limit and then generating a span, so it seems very unlikely that another token could be added to the bucket in that interim. I changed the logic so that we're now calling `_is_allowed()` which will check beforehand if the rate limit is hit without trying to update the rates, if it isn't allowed, then we generate span directly after it, which should hit that limit almost always.